lib/stinger_core/inc/stinger_batch_insert.h

```diff
@@ -0,0 +1,571 @@
+        Iter i = std::lower_bound(begin, end, val, comp);
+
+        if (i != end && !(comp(val, *i)))
+            return i; // found
+        else
+            return end; // not found
+    }
+
+    // Find the first pending update with destination 'dest'
+    template<class use_dest>
+    static iterator
+    find_updates(iterator begin, iterator end, int64_t neighbor)
+    {
+        // Create a dummy update object with the neighbor we are looking for
+        update key;
+        use_dest::set(key, neighbor);
+        // Find the updates for this neighbor
+        iterator pos = binary_find(begin, end, key, use_dest::compare);
+        // If the first update is not pending, we have already done all the updates for this neighbor
+        if (pos != end && adapter::get_result(*pos) == PENDING) return pos;
+        else return end;
+    }
+
+    // Keeps track of the next pending update to insert into the graph
+    class next_update_tracker
+    {
+    protected:
+        iterator pos;
+        iterator end;
+    public:
+        next_update_tracker(iterator begin, iterator end)
+        : pos(begin), end(end) {}
+
+        iterator
+        operator () ()
+        {
+            while (pos != end && adapter::get_result(*pos) != PENDING) { ++pos; }
+            return pos;
+        }
+    };
+
+    /*
+     * Arguments:
+     * result - Result code to set on first update for a neighbor. Result of duplicate updates will always be EDGE_UPDATED.
+     * create - Are we inserting into an empty slot, or just updating an existing slot?
+     * pos - pointer to first update
+     * updates_end - pointer to end of update list
+     * G - pointer to STINGER
+     * eb - pointer to edge block
+     * e - edge index within block
+     * operation - EDGE_WEIGHT_SET or EDGE_WEIGHT_INCR
+     */
+    template<int64_t direction, class use_dest>
+    static void
+    do_edge_updates(int64_t result, bool create, iterator pos, iterator updates_end, stinger_t * G, stinger_eb *eb, size_t e, int64_t operation)
+    {
+        // 'pos' points to the first update for this edge slot; there may be more than one, or none if it equals 'updates_end'
+        // Keep incrementing the iterator until we reach an update with a different neighbor
+        for (iterator u = pos; u != updates_end && use_dest::get(*u) == use_dest::get(*pos); ++u) {
+            int64_t neighbor = use_dest::get(*pos);
+            int64_t weight = adapter::get_weight(*u);
+            int64_t time = adapter::get_time(*u);
+            if (u == pos) {
+                adapter::set_result(*u, result);
+                update_edge_data_and_direction (G, eb, e, neighbor, weight, time, direction, create ? EDGE_WEIGHT_SET : operation);
+            } else {
+                adapter::set_result(*u, EDGE_UPDATED);
+                update_edge_data_and_direction (G, eb, e, neighbor, weight, time, direction, operation);
+            }
```

Comment:
This is already inside a larger parallel for, so it doesn't make sense to parallelize again. 

An edge block consists of a header followed by a fixed-size array (default 14) of edges. I called it an edge slot because its really just a place to store an edge; if we delete the edge or haven't allocated one yet, it's just an empty space.